src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+        /// </summary>
+        protected void SetUnreachable()
         {
-            VisitStatement(tryBlock);
+            this.State = UnreachableState();
         }
 
-        protected virtual void VisitCatchBlock(BoundCatchBlock catchBlock, ref TLocalState finallyState)
+        protected void VisitLvalue(BoundExpression node)
         {
-            if (catchBlock.ExceptionSourceOpt != null)
+            if (_trackRegions && node == this.firstInRegion && this.regionPlace == RegionPlace.Before) EnterRegion();
+            switch (node?.Kind)
             {
-                VisitLvalue(catchBlock.ExceptionSourceOpt);
+                case BoundKind.Parameter:
+                    VisitLvalueParameter((BoundParameter)node);
+                    break;
+
+                case BoundKind.Local:
+                    VisitLvalue((BoundLocal)node);
+                    break;
+
+                case BoundKind.ThisReference:
+                case BoundKind.BaseReference:
+                    // no need for it to be previously assigned: it is on the left.
+                    break;
+
+                case BoundKind.PropertyAccess:
+                    var access = (BoundPropertyAccess)node;
+
+                    if (Binder.AccessingAutoPropertyFromConstructor(access, _member))
+                    {
+                        var backingField = (access.PropertySymbol as SourcePropertySymbol)?.BackingField;
+                        if (backingField != null)
+                        {
+                            VisitFieldAccessInternal(access.ReceiverOpt, backingField);
+                            break;
+                        }
+                    }
+
+                    goto default;
+
+                case BoundKind.FieldAccess:
+                    {
+                        BoundFieldAccess node1 = (BoundFieldAccess)node;
+                        VisitFieldAccessInternal(node1.ReceiverOpt, node1.FieldSymbol);
+                        break;
+                    }
+
+                case BoundKind.EventAccess:
+                    {
+                        BoundEventAccess node1 = (BoundEventAccess)node;
+                        VisitFieldAccessInternal(node1.ReceiverOpt, node1.EventSymbol.AssociatedField);
+                        break;
+                    }
+
+                case BoundKind.TupleLiteral:
+                    ((BoundTupleExpression)node).VisitAllElements((x, self) => self.VisitLvalue(x), this);
+                    break;
+
+                case BoundKind.SuppressNullableWarningExpression:
+                    VisitLvalue(((BoundSuppressNullableWarningExpression)node).Expression);
+                    break;
+
+                default:
+                    VisitRvalue(node);
+                    break;
             }
 
-            if (catchBlock.ExceptionFilterOpt != null)
+            if (_trackRegions && node == this.lastInRegion && this.regionPlace == RegionPlace.Inside) LeaveRegion();
```

Comment:
IMO, it would be much cleaner to have an outer method and an inner method.  The outer method just concerns itself with teh region handling code, and then it calls into the inner method.  The inner method can just do whatever it want, including returning when convenient.  The concern i have with this code is that it would be easy to accidently early return and leave the region-tracking code in an incorrect state.